util/semaphore-group/semaphore_group.go

```diff
@@ -0,0 +1,39 @@
+package semaphoregroup
+
+import (
+	"sync"
+)
+
+// SemaphoreGroup is a custom type that combines sync.WaitGroup and a semaphore.
+type SemaphoreGroup struct {
+	waitGroup sync.WaitGroup
+	semaphore chan struct{}
+}
+
+// NewSemaphoreGroup creates a new SemaphoreGroup with the specified semaphore limit.
+func NewSemaphoreGroup(limit int) *SemaphoreGroup {
+	return &SemaphoreGroup{
+		semaphore: make(chan struct{}, limit),
+	}
+}
+
+// Add increments the internal WaitGroup counter and acquires a semaphore slot.
+func (sg *SemaphoreGroup) Add() {
+	sg.waitGroup.Add(1)
+
+	// Acquire semaphore slot
+	sg.semaphore <- struct{}{}
+}
+
+// Done decrements the internal WaitGroup counter and releases a semaphore slot.
+func (sg *SemaphoreGroup) Done() {
+	sg.waitGroup.Done()
```

Comment:
If we run once time the `Add()` and run multiple times the `Done()` then it will cause an exception. If we use exported functions of this struct from multiple go routines then this part is dangerous.

Reply:
any suggestions?

Reply:
It carries the exact same risk as using a plain sync.WaitGroup, as long as the usage guarantees 1:1 Add and Done usage it should be fine, no?

Reply:
I think the solution is simple. Just fully delete the  sg.waitGroup. The exported Wait has been used by the unit test only, but we can do it in different way. After all Go routine exited, check the len of the channel. It is not nice way to check internal variables in unit test but the code will be simpler and less risky to use.

Reply:
@mohamed-essam I think this is Go philosophy question. The panic can force the developer to write, thread-safe code. I think returning with an error and handling it in the upper layer is a more defensive approach in a production-ready code. If this code snipped would be placed in internal package or it is not exported then the panic is okay from my point of view.